### PR TITLE
Updated table nodes to support soft deletes using the new field

### PR DESCRIPTION
**Pull request by jvelilla**
'deleted_at' as Datetime and give us free metadata.
Updated Sqlite builder to test different scenarios for users and roles.
Updated NODE_FORM_RESPONSE.edit_form feature to add a delete operation iff
there is a node ie node id >0 and the current user has delete permission on it.
Updated NODE_HANDLER.do_post to handle the operation "DELETE".
Updated queries to retrieve nodes filter by no logical deleted rows (ie. deleted_at is NULL).
Updated CMS_USER_API.has_permissions. (authenticated_user_role seems to generic).

### CODE REVIEW

File: examples/demo/site/scripts/node.sql [hunk 1]
```diff
@@ -11,7 +11,8 @@ CREATE TABLE "nodes"(
   "author" INTEGER,
   "publish" DATETIME,
   "created" DATETIME NOT NULL,
-  "changed" DATETIME NOT NULL
+  "changed" DATETIME NOT NULL,
+  "deleted_at" DATETIME NULL
```
Comment (jocelyn) [inline examples/demo/site/scripts/node.sql]: I think we should adopt a different design here.
And go for adding a "status"
and then, "deleted" would be   status = deleted_id  , and deleted_at .. would be "changed" date.

File: library/persistence/sqlite/src/cms_storage_sqlite_builder.e [hunk 2]
```diff
@@ -65,10 +60,40 @@ feature -- Factory
 			create r.make ("authenticated")
 			r.add_permission ("create page")
 			r.add_permission ("edit page")
+			r.add_permission ("delete page")
 			a_storage.save_user_role (r)
 
-				-- Test custom value
```
Comment (jocelyn) [inline library/persistence/sqlite/src/cms_storage_sqlite_builder.e]: I know I left two line of tests for set_custom_value, but this was an error from me, those classes should not contain any "test" purpose data/code.

We should really move them to autotest classes.

Here, we could keep  users + roles initialization .. as they are required.

we could also move that to the USER_MODULE implementation.

File: modules/node/persistence/cms_node_storage_sql.e [hunk 3]
```diff
@@ -260,9 +263,9 @@ feature -- Helpers
 
 feature {NONE} -- Queries
 
-	sql_select_nodes_count: STRING = "SELECT count(*) from Nodes;"
+	sql_select_nodes_count: STRING = "SELECT count(*) from Nodes where deleted_at IS NULL;"
```
Comment (jocelyn) [inline modules/node/persistence/cms_node_storage_sql.e]: please always put SQL keyword as uppercase, this is then much easier to quickly read a SQL statement.

File: src/service/user/cms_user_api.e [hunk 4]
```diff
@@ -49,7 +49,8 @@ feature -- Status report
 				if is_admin_user (a_user) then
 					Result := True
 				else
-					Result := user_role_has_permission (authenticated_user_role, a_permission)
+					fixme ("Check how to handle this predefined role")
+						-- Result := user_role_has_permission (authenticated_user_role, a_permission)
```
Comment (jocelyn) [inline src/service/user/cms_user_api.e]: Please revert this change, there is really an "authenticated" user role ... in fact, as soon as a user is signed in, it has the authenticated user role.
This gives the possibility to set permission for all authenticated users.

File: modules/node/content/cms_node.e [hunk 5]
```diff
@@ -211,6 +217,41 @@ feature -- Element change
 			auther_set: author = u
 		end
 
+	mark_not_published
+			-- Set status to not_published
+		do
+			set_status ({CMS_NODE_CONSTANTS}.not_published)
```
Comment (jocelyn) [inline modules/node/content/cms_node.e]: we could use a {CMS_NODE_API}.status_not_published: INTEGER = 1
without adding a new class.

File: modules/node/content/cms_node.e [hunk 6]
```diff
@@ -78,6 +80,10 @@ feature -- Access
 		deferred
 		end
 
+	status: INTEGER
+			-- Associated status for the current node
+			-- [{1,Not_Published}, {2, Published}, {3, Trash}]
```
Comment (jocelyn) [inline modules/node/content/cms_node.e]: maybe reference the CMS_NODE_CONSTANTS interface or similar.

We could also start at '0' for Not Published, having the default value = 0 as being the default status as not published

File: modules/node/content/cms_node.e [hunk 7]
```diff
@@ -211,6 +217,41 @@ feature -- Element change
...
+
+	mark_published
+			-- Set status to published
+		do
+			set_status ({CMS_NODE_CONSTANTS}.published)
+		ensure
+			status_published: status = {CMS_NODE_CONSTANTS}.published
+		end
+
+	mark_trash
```
Comment (jocelyn) [inline modules/node/content/cms_node.e]: rename  mark_trashed  to follow mark_(not_)published feature name.

File: modules/node/content/cms_node.e [hunk 8]
```diff
@@ -211,6 +217,41 @@ feature -- Element change
...
+			set_status ({CMS_NODE_CONSTANTS}.trash)
+		ensure
+			status_trash: status = {CMS_NODE_CONSTANTS}.trash
+		end
+
+
+feature {NONE} -- Implementation
+
+	set_status (a_status: like status)
+			-- Assign `status' with `a_status'
```
Comment (jocelyn) [inline modules/node/content/cms_node.e]: missing, final dot.
Maybe add comment referencing the {CMS_NODE_CONSTANTS} or similar interface

File: modules/node/content/cms_node_constants.e [hunk 9]
```diff
@@ -0,0 +1,17 @@
+note
+	description: "Node Status Not-Published, Published and Trash"
+	date: "$Date$"
+	revision: "$Revision$"
+
+class
+	CMS_NODE_CONSTANTS
+
+Feature
```
Comment (jocelyn) [inline modules/node/content/cms_node_constants.e]: missing feature clause comment.

File: modules/node/content/cms_node_constants.e [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+
+	Not_published: INTEGER = 1
+
+	Published: INTEGER = 2
+
+	Trash: INTEGER = 3
```
Comment (jocelyn) [inline modules/node/content/cms_node_constants.e]: Renamed  "trashed" to follow other (not_)published feature name.

File: modules/node/handler/node_form_response.e [hunk 11]
```diff
@@ -228,6 +228,12 @@ feature -- Form
 			ts.set_default_value ("Preview")
 			f.extend (ts)
 
+			if a_node /= Void and then a_node.id > 0 and then has_permission ("delete " + a_name) then
```
Comment (jocelyn) [inline modules/node/handler/node_form_response.e]: we may need to distinguish

"delete own " + a_name
and 
"delete any " + a_name

File: modules/node/handler/node_form_response.e [hunk 12]
```diff
@@ -228,6 +228,12 @@ feature -- Form
...
+				create ts.make ("op")
+				ts.set_default_value ("Delete")
```
Comment (jocelyn) [inline modules/node/handler/node_form_response.e]: note for later, for now, we don't do anything for i18n, but at some point, we'll need to find solution, and probably we'lll have to do

ts.set_default_value ("Delete")
with

ts.set_default_value (i18n ("Delete"))
i18n or other name such as "translated" or "translation"

File: modules/node/content/cms_node.e (hunk 5 above)
Comment (jvelilla) [inline modules/node/content/cms_node.e]: I agree

File: modules/node/persistence/cms_node_storage_sql.e [hunk 13]
```diff
@@ -260,24 +265,27 @@ feature -- Helpers
 
 feature {NONE} -- Queries
 
-	sql_select_nodes_count: STRING = "SELECT count(*) from Nodes;"
+	sql_select_nodes_count: STRING = "SELECT count(*) FROM Nodes WHERE status != 3;"
+			-- Nodes count (Published and not Published)
```
Comment (jocelyn) [inline modules/node/persistence/cms_node_storage_sql.e]: Add in comment
{CMS_NODE_CONSTANTS}.trashed = 3

in order to explain why this is 3

File: modules/node/content/cms_node.e (hunk 6 above)
Comment (jvelilla) [inline modules/node/content/cms_node.e]: Agree, and I will do the change to start with 0.

File: modules/node/handler/node_form_response.e (hunk 12 above)
Comment (jvelilla) [inline modules/node/handler/node_form_response.e]: Ok I will add a TODO

File: modules/node/persistence/cms_node_storage_sql.e [hunk 14]
```diff
@@ -335,6 +343,16 @@ feature {NONE} -- Implementation
...
 					Result.set_modification_date (l_modif_date)
 				end
+				if attached sql_read_integer_32 (12) as l_status then
+					inspect l_status
+					when {CMS_NODE_CONSTANTS}.not_published then
+						Result.mark_not_published
+					when {CMS_NODE_CONSTANTS}.published then
+						Result.mark_published
+					when {CMS_NODE_CONSTANTS}.trash then
+						Result.mark_trash
```
Comment (jocelyn) [inline modules/node/persistence/cms_node_storage_sql.e]: add a default "else" part to avoid crash if l_status has wrong value.
and also, if a module introduce a new status, this code will be broken.
Maybe we should not add too much knowledge here, and simply do a set_status (l_status)

File: modules/node/handler/node_form_response.e (hunk 12 above)
Comment (jocelyn) [inline modules/node/handler/node_form_response.e]: This is a general comment that apply to the other classes as well.
but ok add a TODO, or FIXME .. so that we don't forget

File: library/persistence/sqlite/src/cms_storage_sqlite_builder.e [hunk 15]
```diff
@@ -48,27 +48,58 @@ feature -- Factory
 		local
 			u: CMS_USER
 			r: CMS_USER_ROLE
+			l: LIST[CMS_USER_ROLE]
```
Comment (jocelyn) [inline library/persistence/sqlite/src/cms_storage_sqlite_builder.e]: missing space between "LIST" and first bracket "["

File: modules/node/cms_node_api.e [hunk 16]
```diff
@@ -249,6 +249,26 @@ feature -- Access: Node
 			end
 		end
 
+	user_is_node_owner (u: READABLE_STRING_32; nid: INTEGER_64): BOOLEAN
```
Comment (jocelyn) [inline modules/node/cms_node_api.e]: the terminology used by this ROC CMS is "author"
thus, rename user_is_node_owner by

user_is_node_author
and probably, we can just name it "is_node_author (u: ...)"

As well, we should review all query names, and see when we use nid, or node object, similar uid and user object.

File: modules/node/cms_node_api.e [hunk 17]
```diff
@@ -249,6 +249,26 @@ feature -- Access: Node
...
+			end
+		end
+
+feature -- Permission Scope: Node
+
+	permission_scope (u: detachable READABLE_STRING_32; nid: INTEGER_64): STRING
+			-- Result 'own' if the user `u' is the owner of the node `nid', in other case
+			-- `any'.
+		do
+			Result := "any"
```
Comment (jocelyn) [inline modules/node/cms_node_api.e]: What if a role has   "delete any page" enabled, and "delete own page" disabled.
this would be an error made by the maintainer while configuring the permissions.

But this may occur.
So I don't know if this  "any" or "own" is good or not.

File: modules/node/cms_node_api.e [hunk 18]
```diff
@@ -279,6 +299,18 @@ feature -- Change: Node
...
+
+feature -- Node status
+
+	Not_published: INTEGER = 1
+			-- The node is not published.
+
+	Published: INTEGER = 2
+			-- The node is published.
+
+	Trashed: INTEGER = 3
```
Comment (jocelyn) [inline modules/node/cms_node_api.e]: Maybe we could use  -1

not_published: 0   (as default)
published: 1
trashed: -1

-1 sound a standard status to say ... trashed.

to be discussed.

File: modules/node/content/cms_node.e [hunk 19]
```diff
@@ -78,6 +80,10 @@ feature -- Access
 		deferred
 		end
 
+	status: INTEGER
+			-- Associated status for the current node.
+			-- [{0,Not_Published}, {1, Published}, {2, Trash}]
```
Comment (jocelyn) [inline modules/node/content/cms_node.e]: the comment does not use the same value as in CMS_NODE_API
I would just write

[{CMS_NODE_API}.Not_Published}, {CMS_NODE_API}.Published, {CMS_NODE_API}.Trashed]

File: modules/node/content/cms_node.e [hunk 20]
```diff
@@ -211,6 +217,41 @@ feature -- Element change
...
+
+	mark_published
+			-- Set status to published.
+		do
+			set_status ({CMS_NODE_API}.published)
+		ensure
+			status_published: status = {CMS_NODE_API}.published
+		end
+
+	mark_trash
```
Comment (jocelyn) [inline modules/node/content/cms_node.e]: rename `mark_trashed'

**Comment by jocelyn** [comment]
merged manually.
